### PR TITLE
add: effective degrees of freedom for penalized csplines

### PR DESCRIPTION
**Pull request by hlruh**
Adds option to compute the effective degrees of freedom. The effective degrees of freedom replace the classic degrees of freedom in the computation of the AIC and BIC information criteria for the pspline (Eilers & Marx, 1996). Computing the AIC or BIC allows to optimize the pspline penalty.

### CODE REVIEW

File: src/cspline.py [hunk 1]
```diff
@@ -409,8 +409,7 @@ def __call__(self, x=None):
             v_f += B[k] * self.cov[kk+k,kk+j] * B[j]
       return v_f.reshape(kk.shape)
 
-
-def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu=None, e_mu=None, nat=True, retfit=False, var=False, e_yk=False, cov=False, plot=False, c=True):
+def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu=None, e_mu=None, nat=True, retfit=False, var=False, e_yk=False, cov=False, ic= False, plot=False, c=True):
```
Comment (mzechmeister) [inline src/cspline.py]: Remove space after `ic= `

File: src/cspline.py [hunk 2]
```diff
@@ -448,13 +447,15 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
    cov : boolean
       If false band matrices are used for efficient solution of equation system with band solver.
       If true covariances are estimated using matrix inversion.
+   ic : float, float, float
```
Comment (mzechmeister) [inline src/cspline.py]: type is boolean

File: src/cspline.py [hunk 3]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+            # design matrix
+            B = bspline2((x-x.min())/(x.max()-x.min())*K, K, D=3)
+                
+            # penalty
+            n = B.shape[1]
+            D = np.diff(np.eye(n), n=pord).T
+            lamDTD = lam*np.dot(D.T, D)
+                
+            # Compute the hat matrix for EDF
+            BT = B.T
```
Comment (mzechmeister) [inline src/cspline.py]: Is a new variable `BT` needed?

File: src/cspline.py [hunk 4]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+            BTWB = np.dot(np.dot(BT, W), B)
```
Comment (mzechmeister) [inline src/cspline.py]: Swap lines and so BTW can be reused.

File: src/cspline.py [hunk 5]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+        if 1:
+            # band matrices version -> faster
+
+            # Compute inverse for effective degrees of freedom (edf)
+            BTWB_lamDTDinv = solveh_banded(BTWB_lamDTDb, np.eye(np.shape(BTWB_lamDTDb)[1]), lower=True)
+
+            # effective degrees of freedom (edf)
+            edf = loop_sum_edf(BTWB_lamDTDinv, G, kk, w)
+            
+            # Compute BIC & AIC
```
Comment (mzechmeister) [inline src/cspline.py]: move outside if-else so that the code block needs not to be duplicated

File: src/cspline.py [hunk 6]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+        A 1D array of weights (size Ndat) for each column in G.
+
+    Returns:
+    --------
+    float
+        Computed effective degrees of freedom.
+    """
+       
+   # get matrix dimensions
+   Nknot, Ndat = BTWB_lamDTDinv.shape[0], G.shape[1]
```
Comment (mzechmeister) [inline src/cspline.py]: Use two lines.

File: src/cspline.py [hunk 7]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+   # precompute for speed
+   Gw = G * w[np.newaxis,:] 
+   GTGw = G[np.newaxis,:,:] * Gw[:,np.newaxis,:]
+   
+   # knot indices
+   kn, ln = np.mgrid[:4,:4]
+   kn = kn.ravel()
+   ln = ln.ravel()
+
+   edf=0
```
Comment (mzechmeister) [inline src/cspline.py]: space around `=`

File: src/cspline.py [hunk 8]
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
...
+   for i in range(Ndat):
+   # loop over each data points
+      kki = kk[i]
+      if kki < Nknot-4:
+         edf += np.sum(BTWB_lamDTDinv[kki+kn, kki+ln] * GTGw[kn,ln,i])
+      else:
+         # need to check if elements are valid for the last elements of GTGw
+         valid = np.flatnonzero((kki+kn < Nknot) & (kki+ln < Nknot))
+         if valid.size > 0:
+            # if elements are valid
```
Comment (mzechmeister) [inline src/cspline.py]: Remove comment.

File: src/cspline.py (hunk 1 above)
Comment (hlruh) [reply src/cspline.py]: replaced by edf

File: src/cspline.py (hunk 2 above)
Comment (hlruh) [reply src/cspline.py]: done

File: src/cspline.py (hunk 3 above)
Comment (hlruh) [reply src/cspline.py]: not needed. removed.

File: src/cspline.py (hunk 4 above)
Comment (hlruh) [reply src/cspline.py]: done.

File: src/cspline.py (hunk 5 above)
Comment (hlruh) [reply src/cspline.py]: done.

File: src/cspline.py (hunk 6 above)
Comment (hlruh) [reply src/cspline.py]: done.

File: src/cspline.py (hunk 7 above)
Comment (hlruh) [reply src/cspline.py]: done.

File: src/cspline.py (hunk 8 above)
Comment (hlruh) [reply src/cspline.py]: done.

**Pull request merged**